src/Components/Components/src/Microsoft.AspNetCore.Components.csproj

```diff
@@ -62,4 +57,8 @@
     <NuspecProperty Include="PackageThirdPartyNoticesFile=$(PackageThirdPartyNoticesFile)" />
   </ItemGroup>
 
+  <ItemGroup>
+    <ProjectReference Include="..\..\..\JSInterop\Microsoft.JSInterop\src\Microsoft.JSInterop.csproj" />
```

Comment:
Ah, unfortunately I didn't realise this would be needed before.

We want to avoid referencing JS interop concepts directly from the `.Components` package because that's meant to be independent of web technology. There are other forms of component rendering that don't involve web browsers or the concept of JavaScript at all.

I know it's quirky that `ElementReference` is a concept inside `.Components`, but the intention is that although for HTML rendering an "element" maps to an HTML element, for other forms of rendering an "element" maps to whatever is the primitive unit of UI in that technology. For example, in our Xamarin Forms renderer, the "element" concept corresponds to an underlying native Xamarin Forms control, and in this form of rendering, there is no JavaScript and hence no JS interop.

I'm not sure if this undermines the whole idea of eliminating the `IJSRuntime` parameter to `focus`. Maybe there might still be some way of doing it, for example by storing the `IServiceProvider` on `ElementReference` so that the `Focus` extension method (which is in the `.Web` project, which is allowed to have a concept of JS interop) can dynamically look up the `IJSRuntime` from the service provider. For the avoidance of doubt, we don't want to expose the `Renderer` instance to application code.

Maybe there's another cleverer solution still, not sure!

Reply:
![image](https://user-images.githubusercontent.com/174281/85636068-5642e280-b634-11ea-8259-60bace35d900.png)

Reply:
I'm not really sure what's it there for though.

Reply:
I didn’t realise the reference was there. I guess it would be difficult to remove now. 

But in any case we shouldn’t assume in this project that the service provider will have any IJSRuntime. It won’t in the case of Mobile Blazor Bindings.

Reply:
Would be interesting to understand how the reference got there. It might be an unintended consequence of the SDK setup or something. It’s too late here for me to dig into it right now, but I wonder if this project really makes use of that reference at all.

Reply:
Building on @MackinnonBuck's suggestion, we could avoid exposing this as part of Component's API surface. It might make it easier to break this if we'd like to before 5.0:

```
internal object? JSRuntime { get; set; }

Components.Web
{
  public static IJSRuntime GetJSRuntime(this ElementReference reference) => reference.JSRuntime ?? throw new PlatformNotSupportedException(..);
}

Reply:
If we’re happy to add an InternalsVisibleTo then yes, I agree that avoids any impact on public API surface.

Reply:
But actually I’m not sure how the renderer is meant to supply that without knowing about the IJSRuntime.

Reply:
Let's do that. We could consider making a breaking change for 5.0 to move the package reference to Components.Web. I think all we really need is the type of IJSRuntime to lookup from DI which does not require the reference.

Reply:
Alternatively, let's get this change in and we can figure it out in a follow up. The package reference exists, and we're only making it slightly worse 👍

Reply:
I've been looking into the suggestion of storing the `IServiceProvider` inside the `ElementReference` and I don't think this can easily be done since each `ElementReferenceJsonConverter` would need to be initialized with an `IServiceProvider` in order to deserialize `ElementReference`s. And it isn't always possible to provide one (e.g. in the case of `DefaultWebAsssemblyJSRuntime`).

Reply:
So just to clarify, are we still adding `ElementReference.JSRuntime` to the public API, are are we hiding it?

Reply:
I just checked this with Pranav and he and I think the `IServiceProvider` option is likely the cleanest one. @pranavkm can you provide pointers on how to get access to the service provider from inside `DefaultWebAssemblyJSRuntime`?